### ISSUE TITLE
maxConnections on outbound clusters should scale according to number of servers the clusters contain

### ISSUE THREAD

**Issue by yann-soubeyrand** [istio/istio#28666]
**Bug description**

Let's consider an application composed of 10 Unicorn servers behind two Varnish cache proxies. Due to how Unicorn works, we want to limit the number of connections to each server to, say, 8. We do this by setting trafficPolicy.connectionPool.tcp.maxConnections to 8 in the destination rule associated with the service provided by the Unicorn servers. However, doing so leads Varnish proxies to open a total of 8 connections each to the Unicorn servers (not 8 connections to each Unicorn servers).

[ ] Docs
[ ] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Expected behavior**

Ideally, each Unicorn server should have 8 connections open, balanced between the Varnish proxies. However, I'm not sure how this could be realized in a generic way (this example application is a really simple one with a single consumer service). A good compromise would be for each Varnish proxy to be able to open 8 connections to each Unicorn servers. This way, even if one of the proxies get down, the full servers capacity can still be exploited.

**Steps to reproduce the bug**

1. Set trafficPolicy.connectionPool.tcp.maxConnections to 8 on the destination rule of the Unicorn service.
2. Observe that maxConnections is set to 8 on the Envoy inbound cluster of the Unicorn servers, which is expected.
3. Observe that maxConnections is set to 8 on the Envoy outbound cluster of the Varnish servers, which causes the problem. I'm not sure what would happen if this value was unlimited, however, I expect it to be set to no less than 80 (8 connections by server × 10 servers).

A similar issue happens for trafficPolicy.connectionPool.http.http1MaxPendingRequests: setting it to 100 affects the Envoy inbound cluster and outbound cluster whereas I would expect only the inbound cluster to be affected.

**Version (include the output of `istioctl version --remote` and `kubectl version --short` and `helm version` if you used Helm)**

```
❯ istioctl version --remote
client version: 1.7.1
control plane version: 1.6.12-92cb89444cfa76e5a1363aaa255e1e777abf80d4
data plane version: 1.6.12 (30 proxies)
```

```
Client Version: v1.16.12
Server Version: v1.17.9-eks-4c6976
```

```
helm version
version.BuildInfo{Version:"v3.3.4", GitCommit:"a61ce5633af99708171414353ed49547cf05013d", GitTreeState:"clean", GoVersion:"go1.14.9"}
```

**How was Istio installed?**

Using Istio Operator and an IstioOperator custom resource.

**Environment where bug was observed (cloud vendor, OS, etc)**

AWS EKS 1.17.

**Comment by yann-soubeyrand** [comment]
CC @hzxuzhonghu following our discussion on Slack (https://istio.slack.com/archives/C38CF1PEC/p1603826093253600).

**Comment by rdineshshetty** [comment]
> CC @hzxuzhonghu following our discussion on Slack (https://istio.slack.com/archives/C38CF1PEC/p1603826093253600).

I dont have access to slack to follow the discussion, am also facing similar problem in istio 1.7.4.. would like to restrict inbound max connections to each pod and scale whenever the threshold is reached..

**Comment by rdineshshetty** [comment]
> > CC @hzxuzhonghu following our discussion on Slack (https://istio.slack.com/archives/C38CF1PEC/p1603826093253600).
> 
> I dont have access to slack to follow the discussion, am also facing similar problem in istio 1.7.4.. would like to restrict inbound max connections to each pod and scale whenever the threshold is reached..

@yann-soubeyrand @hzxuzhonghu

**Comment by yann-soubeyrand** [comment]
@rdineshshetty there was nothing in the Slack discussion which is not in this report, I was just linking it to @hzxuzhonghu so that he knew this issue is related to the discussion we had. I’m not a maintainer so I cannot re-open this issue.

**Comment by rdineshshetty** [comment]
> > > CC @hzxuzhonghu following our discussion on Slack (https://istio.slack.com/archives/C38CF1PEC/p1603826093253600).
> > 
> > 
> > I dont have access to slack to follow the discussion, am also facing similar problem in istio 1.7.4.. would like to restrict inbound max connections to each pod and scale whenever the threshold is reached..
> 
> @yann-soubeyrand @hzxuzhonghu

@howardjohn @rshriram any suggestion/work-around to handle this scenario

**Comment by rdineshshetty** [comment]
> @rdineshshetty there was nothing in the Slack discussion which is not in this report, I was just linking it to @hzxuzhonghu so that he knew this issue is related to the discussion we had. I’m not a maintainer so I cannot re-open this issue.

thanks, somehow i missed your reply @yann-soubeyrand ..  wanted to know, whether you got any solution or workaround for the scenario you have mentioned..

**Comment by yann-soubeyrand** [comment]
@rdineshshetty sorry, I don’t really have a workaround…

**Comment by ankitdiscovery** [comment]
@yann-soubeyrand  any workaround solution for this ?

**Comment by hzxuzhonghu** [comment]
It is hard to do this in istio. Currently istio can do incremental eds push only during workload scale up and down, but we do not push CDS now. And I do not think we can change the API semantics to support this.

However IMO, we can use connection limit filter to limit in the inbound listener. And we do not set the limit with destination rule connection pool

https://www.envoyproxy.io/docs/envoy/latest/configuration/listeners/network_filters/connection_limit_filter#connection-limit-filter

**Comment by ankitdiscovery** [comment]
@yann-soubeyrand 

i have istio service mesh with envoy running as sidecar, a destination service with below configuration in destination rule is deployed. mainly i was trying to check ciruitbreaker beaviour of envoy/istio.
```
     trafficPolicy:
   connectionPool:
     http:
       http1MaxPendingRequests: 1
     tcp:
       maxConnections: 5
   ```
       
my client proxy is sending request to this and i see a same maxconnection limit getting applied to client side also. if a scale destination replicas to 4, would be my destination service work with total 20 connections (5 connections each pod ) ? or 5 strict connections would be shared between all 4 pods ?

context:- i see after increasing the destination replicas i get less 503 with reponse flag UO.

**Comment by hamishforbes** [comment]
> 



> However IMO, we can use connection limit filter to limit in the inbound listener. And we do not set the limit with destination rule connection pool
> 
> https://www.envoyproxy.io/docs/envoy/latest/configuration/listeners/network_filters/connection_limit_filter#connection-limit-filter

Has anyone managed to get this to work correctly?

I have a case where my server app will essentially fall over and die with >N connections.
DestinationRule is very unhelpful here, even if I do some hackery to set the maxConnections based on the number of replicas on both the client and server deployments, it only takes a rolling restart to throw those numbers out

I really want to be able to put a concrete limit on the number of inbound connections to my server app.
A circuit breaker in the inbound cluster is the obvious solution and I'm struggling to get the listener connection filter to work

**Comment by yann-soubeyrand** [comment]
Hi @hzxuzhonghu, am I right in thinking that this issue is kind of fixed in 1.21 by using sidecar inbound connection pool (https://istio.io/latest/docs/reference/config/networking/sidecar/#Sidecar-inbound_connection_pool) instead of destination rule connection pool?
